pkg/metrics/query.go

```diff
@@ -0,0 +1,218 @@
+package metrics
+
+import (
+	"crypto/tls"
+	"crypto/x509"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"strings"
+	"time"
+
+	corev1 "k8s.io/api/core/v1"
+	kapierrs "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/kubernetes"
+
+	"github.com/prometheus/common/model"
+)
+
+const (
+	serviceCAinjectionDataKey             = "service-ca.crt"
+	serviceCAinjectCABundleAnnotationName = "service.beta.openshift.io/inject-cabundle"
+	metricsBaseURL                        = "https://prometheus-k8s.openshift-monitoring.svc:9091"
+)
+
+type PromQuery struct {
+	Client        *kubernetes.Clientset
+	Query         string
+	ConfigMapName string
+	NsName        string
+	URL           string
+	BearerToken   string
+}
+
+func NewPromQuery(client *kubernetes.Clientset, query, cmName, nsName string) (*PromQuery, error) {
+	baseurl, bt, ok := locatePrometheus(client)
+	if !ok {
+		return nil, fmt.Errorf("Could not locate prometheus")
+	}
+	url := fmt.Sprintf("%s/api/v1/query?%s", baseurl, (url.Values{"query": []string{query}}).Encode())
+	return &PromQuery{
+		Client:        client,
+		Query:         query,
+		ConfigMapName: cmName,
+		NsName:        nsName,
+		URL:           url,
+		BearerToken:   bt,
+	}, nil
+}
+
+func locatePrometheus(client *kubernetes.Clientset) (url, bearerToken string, ok bool) {
+	_, err := client.CoreV1().Services("openshift-monitoring").Get("prometheus-k8s", metav1.GetOptions{})
+	if kapierrs.IsNotFound(err) {
+		return "", "", false
+	}
+
+	err = wait.Poll(time.Second*5, time.Second*300, func() (bool, error) {
+		secrets, err := client.CoreV1().Secrets("openshift-monitoring").List(metav1.ListOptions{})
+		if err != nil {
+			return false, fmt.Errorf("could not list secrets in openshift-monitoring namespace")
+		}
+		for _, secret := range secrets.Items {
+			if secret.Type != corev1.SecretTypeServiceAccountToken {
+				continue
+			}
+			if !strings.HasPrefix(secret.Name, "prometheus-") {
+				continue
+			}
+			bearerToken = string(secret.Data[corev1.ServiceAccountTokenKey])
```

Comment:
Is this the only way to get access to prometheus?  I'm wondering if there is an option that doesn't require the same privileges as prometheus itself.

Reply:
this is how it's done in [origin/test/extended/prometheus/prometheus.go](https://github.com/openshift/origin/blob/master/test/extended/prometheus/prometheus.go#L469-#L498)  that's what I based this off of.

Reply:
I've added helpers to run exec curl - this seems to be the only way to exercise the cluster CA

Reply:
never mind, we've determined that running the query against the external prometheus URL, no exec, no CAbundle required is sufficient